src/sync.rs

```diff
@@ -527,6 +531,100 @@ where
     }
 }
 
+/// Executes a closure with a Python critical section held on mutex.
+///
+/// Acquires the per-object lock for the mutex `mutex` that is held
+/// until the closure `f` is finished.
+///
+/// This is structurally equivalent to the use of the paired
+/// Py_BEGIN_CRITICAL_SECTION_MUTEX and Py_END_CRITICAL_SECTION C-API macros.
+///
+/// A no-op on GIL-enabled builds, where the critical section API is exposed as
+/// a no-op by the Python C API.
+///
+/// Provides weaker locking guarantees than traditional locks, but can in some
+/// cases be used to provide guarantees similar to the GIL without the risk of
+/// deadlocks associated with traditional locks.
+///
+/// This variant with a mutex as an argument is particularly useful when paired
+/// with a static `PyMutex`. This can be used in situations where a `PyMutex` would
+/// deadlock with the interpreter.
+///
+/// Many CPython C API functions do not acquire the per-object lock on objects
+/// passed to Python. You should not expect critical sections applied to
+/// built-in types to prevent concurrent modification. This API is most useful
+/// for user-defined types with full control over how the internal state for the
+/// type is managed.
+///
+/// Only available on Python 3.14 and newer.
+#[cfg(Py_3_14)]
+#[cfg_attr(not(Py_GIL_DISABLED), allow(unused_variables))]
+pub fn with_critical_section_mutex<F, R, T>(mutex: &mut PyMutex<T>, f: F) -> R
+where
+    F: FnOnce(&mut T) -> R,
```

Comment:
Hmm. I think because critical sections don't actually guarantee exclusivity for the whole closure due to possibility of nesting and release due to e.g. GC, we have to still make users be `unsafe` when accessing the inner content and uphold the access pattern themselves.

I also think we don't need to borrow the mutex itself mutably.

Something like the following might work:

```suggestion
pub fn with_critical_section_mutex<F, R, T>(mutex: &PyMutex<T>, f: F) -> R
where
    F: FnOnce(&UnsafeCell<T>) -> R,
```

we might also want to package the `&UnsafeCell<T>` as something a bit nicer? e.g.

```rust
struct InCriticalSection<'a, T>(&'a UnsafeCell<T>);

impl InCriticalSection<'_, T> {
    /// Assert that this critical section is the outermost section and access the inner value.
    ///
    /// # Safety
    /// 
    /// This type cannot be used safely within nested critical sections.
    unsafe fn assert_outermost_section(&self) -> &mut T {
        unsafe { &mut *self.0.get() }
    }
}
```

... on Rust nightly, a custom auto trait would probably have solved this but they seem to be falling out of favour.